include/phasar/PhasarLLVM/DataFlowSolver/IfdsIde/EdgeFunctionUtils.h

```diff
@@ -0,0 +1,320 @@
+
+  ~AllBottom() override = default;
+
+  EdgeFunctionPtrType composeWith(EdgeFunctionPtrType SecondFunction) override {
+    return SecondFunction->isConstant() ? SecondFunction
+                                        : this->shared_from_this();
+  }
+
+  EdgeFunctionPtrType joinWith(EdgeFunctionPtrType /*OtherFunction*/) override {
+    return this->shared_from_this();
+  }
+
+  void print(llvm::raw_ostream &OS,
+             bool /*IsForDebug = false*/) const override {
+    OS << "AllBottom";
+  }
+
+  [[nodiscard]] bool isConstant() const noexcept override { return true; }
+
+  template <typename LL = L,
+            typename = std::enable_if_t<HasJoinLatticeTraits<LL>>>
+  static ByConstRef<EdgeFunctionPtrType> getInstance() {
+    static EdgeFunctionPtrType AllBotFn = std::make_shared<AllBottom>();
+    return AllBotFn;
+  }
+};
+
+template <typename L>
+class AllTop : public detail::AllTopBase<L>,
+               public std::enable_shared_from_this<AllTop<L>> {
+public:
+  using typename EdgeFunction<L>::EdgeFunctionPtrType;
+
+  using detail::AllTopBase<L>::AllTopBase;
+
+  ~AllTop() override = default;
+
+  EdgeFunctionPtrType composeWith(EdgeFunctionPtrType SecondFunction) override {
+    if (SecondFunction == EdgeIdentity<L>::getInstance()) {
+      return this->shared_from_this();
+    }
+    return SecondFunction;
+  }
+
+  EdgeFunctionPtrType joinWith(EdgeFunctionPtrType OtherFunction) override {
+    return OtherFunction;
+  }
+
+  void print(llvm::raw_ostream &OS,
+             [[maybe_unused]] bool IsForDebug = false) const override {
+    OS << "AllTop";
+  }
+
+  [[nodiscard]] bool isConstant() const noexcept override { return true; }
+};
+
+template <typename L>
+auto EdgeIdentity<L>::joinWith(EdgeFunctionPtrType OtherFunction)
+    -> EdgeFunctionPtrType {
+  if ((OtherFunction.get() == this) ||
+      OtherFunction->equal_to(this->shared_from_this())) {
+    return this->shared_from_this();
+  }
+  if (auto *AB = dynamic_cast<AllBottom<L> *>(OtherFunction.get())) {
+    return OtherFunction;
+  }
+  if (auto *AT = dynamic_cast<AllTop<L> *>(OtherFunction.get())) {
+    return this->shared_from_this();
+  }
+  // do not know how to join; hence ask other function to decide on this
+  return OtherFunction->joinWith(this->shared_from_this());
+}
+
+template <typename L, typename Enable>
+auto ConstantEdgeFunction<L, Enable>::joinWith(
+    EdgeFunctionPtrType OtherFunction) -> EdgeFunctionPtrType {
+  if (this == OtherFunction.get() || equal_to(OtherFunction)) {
+    return OtherFunction;
+  }
+
+  if (!OtherFunction->isConstant()) {
+    return AllBottom<L>::getInstance();
+  }
+
+  auto OtherVal = OtherFunction->computeTarget(JoinLatticeTraits<L>::top());
+  auto JoinedVal = JoinLatticeTraits<L>::join(Value, OtherVal);
+
+  if (JoinLatticeTraits<L>::bottom() == JoinedVal) {
+    return AllBottom<L>::getInstance();
+  }
+
+  if (JoinLatticeTraits<L>::top() == JoinedVal) [[unlikely]] {
+    // XXX: Can this ever happen?
```

Comment:
If your not sure about the case, maybe put some assert here, to verify the semantics should it be triggered in a case.